src/test-run/index.js

```diff
@@ -0,0 +1,142 @@
+import path from 'path';
+import Promise from 'pinkie';
+import { Session } from 'testcafe-hammerhead';
+import { createDoneCommand, TYPE as COMMAND_TYPE } from './commands';
+import CLIENT_MESSAGES from './client-messages';
+import delay from '../utils/delay';
+import MESSAGE from '../errors/runtime/message';
+import { GeneralError } from '../errors/runtime';
+
+
+export default class TestRun extends Session {
+    constructor (test, browserConnection, screenshotCapturer, opts) {
+        var uploadsRoot = path.dirname(test.fixture.path);
+
+        super(uploadsRoot);
+
+        this.opts               = opts;
+        this.test               = test;
+        this.browserConnection  = browserConnection;
+        this.screenshotCapturer = screenshotCapturer;
+
+        this.started        = false;
+        this.pendingCommand = null;
+        this.pendingRequest = null;
+        this.pendingJsError = null;
+
+        this.errs = [];
+    }
+
+    _getPayloadScript () {
+        // TODO
+    }
+
+    _getIframePayloadScript () {
+        // TODO
+    }
+
+    async _start () {
+        this.started = true;
+
+        try {
+            await this.test.fn(this);
+        }
+        catch (err) {
+            this._fatalError(err);
+        }
+        finally {
+            this._done();
+        }
+    }
+
+    _fatalError (err) {
+        this.errs.push(err);
+    }
+
+    _done () {
+        if (this.pendingRequest)
+            this._resolvePendingRequest(createDoneCommand());
```

Comment:
It's inconsistent. What if we've nulled pending request just now? Client will never know that we've finished task. The best way will be to run `executeCommand` to use same consistent machinery